File: server/src/services/socketService.js
import WatchRoom from '../models/WatchRoom.model.js';

// Store active socket connections per room
const roomSockets = new Map(); // roomCode -> Set of socket ids
const socketRooms = new Map(); // socketId -> roomCode

export const initializeSocket = (io) => {
    io.on('connection', (socket) => {
        console.log('User connected:', socket.id);

        // Join a watch room
        socket.on('join-room', async (data) => {
            try {
                const { roomCode, userId, userName } = data;
                const upperCode = roomCode.toUpperCase();

                // Leave previous room if any
                const previousRoom = socketRooms.get(socket.id);
                if (previousRoom) {
                    socket.leave(previousRoom);
                    const roomSet = roomSockets.get(previousRoom);
                    if (roomSet) {
                        roomSet.delete(socket.id);
                    }
                }

                // Join new room
                socket.join(upperCode);
                socketRooms.set(socket.id, upperCode);

                if (!roomSockets.has(upperCode)) {
                    roomSockets.set(upperCode, new Set());
                }
                roomSockets.get(upperCode).add(socket.id);

                // Get current room state from DB
                const room = await WatchRoom.findOne({
                    roomCode: upperCode,
                    isActive: true
                });

                if (room) {
                    // Send current state to the joining user
                    socket.emit('sync-state', {
                        currentTime: room.currentTime,
                        isPlaying: room.isPlaying,
                        currentServer: room.currentServer,
                        currentEpisode: room.currentEpisode
                    });

                    // Notify others about new participant
                    socket.to(upperCode).emit('user-joined', {
                        oderId: userId,
                        userName: userName,
                        participantCount: roomSockets.get(upperCode).size
                    });
                }

                console.log(`User ${socket.id} joined room ${upperCode}`);
            } catch (error) {
                console.error('Join room error:', error);
                socket.emit('error', { message: 'Failed to join room' });
            }
        });

        // Leave room
        socket.on('leave-room', (data) => {
            const { roomCode } = data;
            const upperCode = roomCode.toUpperCase();

            socket.leave(upperCode);
            socketRooms.delete(socket.id);

            const roomSet = roomSockets.get(upperCode);
            if (roomSet) {
                roomSet.delete(socket.id);

                // Notify others
                socket.to(upperCode).emit('user-left', {
                    participantCount: roomSet.size
                });
            }

            console.log(`User ${socket.id} left room ${upperCode}`);
        });

        // Video play event
        socket.on('video-play', async (data) => {
            const { roomCode, currentTime, userId } = data;
            const upperCode = roomCode.toUpperCase();

            // Update DB
            await WatchRoom.findOneAndUpdate(
                { roomCode: upperCode, isActive: true },
                { isPlaying: true, currentTime: currentTime }
            );

            // Broadcast to others in room
            socket.to(upperCode).emit('video-play', {
                currentTime: currentTime,
                triggeredBy: userId
            });

            console.log(`Room ${upperCode}: Play at ${currentTime}s`);
        });

        // Video pause event
        socket.on('video-pause', async (data) => {
            const { roomCode, currentTime, userId } = data;
            const upperCode = roomCode.toUpperCase();

            // Update DB
            await WatchRoom.findOneAndUpdate(
                { roomCode: upperCode, isActive: true },
                { isPlaying: false, currentTime: currentTime }
            );

            // Broadcast to others in room
            socket.to(upperCode).emit('video-pause', {
                currentTime: currentTime,
                triggeredBy: userId
            });

            console.log(`Room ${upperCode}: Pause at ${currentTime}s`);
        });

        // Video seek event
        socket.on('video-seek', async (data) => {
            const { roomCode, currentTime, userId } = data;
            const upperCode = roomCode.toUpperCase();

            // Update DB
            await WatchRoom.findOneAndUpdate(
                { roomCode: upperCode, isActive: true },
                { currentTime: currentTime }
            );

            // Broadcast to others in room
            socket.to(upperCode).emit('video-seek', {
                currentTime: currentTime,
                triggeredBy: userId
            });

            console.log(`Room ${upperCode}: Seek to ${currentTime}s`);
        });

        // Episode change event
        socket.on('episode-change', async (data) => {
            const { roomCode, serverIndex, episodeIndex, userId } = data;
            const upperCode = roomCode.toUpperCase();

            // Update DB
            await WatchRoom.findOneAndUpdate(
                { roomCode: upperCode, isActive: true },
                {
                    currentServer: serverIndex,
                    currentEpisode: episodeIndex,
                    currentTime: 0,
                    isPlaying: false
                }
            );

            // Broadcast to others in room
            socket.to(upperCode).emit('episode-change', {
                serverIndex: serverIndex,
                episodeIndex: episodeIndex,
                triggeredBy: userId
            });

            console.log(`Room ${upperCode}: Episode changed to S${serverIndex}E${episodeIndex}`);
        });

        // Request sync from host
        socket.on('sync-request', async (data) => {
            const { roomCode } = data;
            const upperCode = roomCode.toUpperCase();

            const room = await WatchRoom.findOne({
                roomCode: upperCode,
                isActive: true
            });

            if (room) {
                socket.emit('sync-state', {
                    currentTime: room.currentTime,
                    isPlaying: room.isPlaying,
                    currentServer: room.currentServer,
                    currentEpisode: room.currentEpisode
                });
            }
        });

        // Room closed by host
        socket.on('close-room', (data) => {
            const { roomCode } = data;
            const upperCode = roomCode.toUpperCase();

            // Notify all users in room
            io.to(upperCode).emit('room-closed', {
                message: 'Host has closed the room'
            });

            // Clear room data
            roomSockets.delete(upperCode);

            console.log(`Room ${upperCode} closed`);
        });

        // Disconnect
        socket.on('disconnect', () => {
            const roomCode = socketRooms.get(socket.id);

            if (roomCode) {
                const roomSet = roomSockets.get(roomCode);
                if (roomSet) {
                    roomSet.delete(socket.id);

                    // Notify others
                    socket.to(roomCode).emit('user-left', {
                        participantCount: roomSet.size
                    });
                }
                socketRooms.delete(socket.id);
            }

            console.log('User disconnected:', socket.id);
        });
    });
};

export default initializeSocket;

File: lib/services/socket_service.dart
import 'dart:async';

import 'package:socket_io_client/socket_io_client.dart' as io;

import '../models/watch_room_model.dart';
import 'api_config.dart';

/// Singleton service to manage Socket.IO connection for WatchAlong feature
class SocketService {
  static final SocketService _instance = SocketService._internal();
  factory SocketService() => _instance;
  SocketService._internal();

  io.Socket? _socket;
  String? _currentRoomCode;
  String? _userId;

  // Stream controllers for events
  final _onVideoPlayController = StreamController<VideoSyncState>.broadcast();
  final _onVideoPauseController = StreamController<VideoSyncState>.broadcast();
  final _onVideoSeekController = StreamController<VideoSyncState>.broadcast();
  final _onSyncStateController = StreamController<VideoSyncState>.broadcast();
  final _onEpisodeChangeController =
      StreamController<Map<String, dynamic>>.broadcast();
  final _onUserJoinedController =
      StreamController<Map<String, dynamic>>.broadcast();
  final _onUserLeftController =
      StreamController<Map<String, dynamic>>.broadcast();
  final _onRoomClosedController = StreamController<String>.broadcast();
  final _onConnectedController = StreamController<bool>.broadcast();

  // Streams for listening
  Stream<VideoSyncState> get onVideoPlay => _onVideoPlayController.stream;
  Stream<VideoSyncState> get onVideoPause => _onVideoPauseController.stream;
  Stream<VideoSyncState> get onVideoSeek => _onVideoSeekController.stream;
  Stream<VideoSyncState> get onSyncState => _onSyncStateController.stream;
  Stream<Map<String, dynamic>> get onEpisodeChange =>
      _onEpisodeChangeController.stream;
  Stream<Map<String, dynamic>> get onUserJoined =>
      _onUserJoinedController.stream;
  Stream<Map<String, dynamic>> get onUserLeft => _onUserLeftController.stream;
  Stream<String> get onRoomClosed => _onRoomClosedController.stream;
  Stream<bool> get onConnected => _onConnectedController.stream;

  bool get isConnected => _socket?.connected ?? false;
  String? get currentRoomCode => _currentRoomCode;

  /// Connect to Socket.IO server
  void connect({String? token}) {
    if (_socket != null && _socket!.connected) {
      print('Socket already connected');
      return;
    }

    _socket = io.io(
      ApiConfig.socketUrl,
      io.OptionBuilder()
          .setTransports(['websocket'])
          .disableAutoConnect()
          .setExtraHeaders(
            token != null ? {'Authorization': 'Bearer $token'} : {},
          )
          .build(),
    );

    _socket!.onConnect((_) {
      print('Socket connected');
      _onConnectedController.add(true);
    });

    _socket!.onDisconnect((_) {
      print('Socket disconnected');
      _onConnectedController.add(false);
    });

    _socket!.onError((error) {
      print('Socket error: $error');
    });

    // Listen to video events
    _socket!.on('video-play', (data) {
      _onVideoPlayController.add(VideoSyncState.fromJson(data));
    });

    _socket!.on('video-pause', (data) {
      _onVideoPauseController.add(VideoSyncState.fromJson(data));
    });

    _socket!.on('video-seek', (data) {
      _onVideoSeekController.add(VideoSyncState.fromJson(data));
    });

    _socket!.on('sync-state', (data) {
      _onSyncStateController.add(VideoSyncState.fromJson(data));
    });

    _socket!.on('episode-change', (data) {
      _onEpisodeChangeController.add(Map<String, dynamic>.from(data));
    });

    // Listen to room events
    _socket!.on('user-joined', (data) {
      _onUserJoinedController.add(Map<String, dynamic>.from(data));
    });

    _socket!.on('user-left', (data) {
      _onUserLeftController.add(Map<String, dynamic>.from(data));
    });

    _socket!.on('room-closed', (data) {
      _onRoomClosedController.add(data['message'] ?? 'Room closed');
      _currentRoomCode = null;
    });

    _socket!.connect();
  }

  /// Disconnect from server
  void disconnect() {
    if (_currentRoomCode != null) {
      leaveRoom(_currentRoomCode!);
    }
    _socket?.disconnect();
    _socket = null;
  }

  /// Join a watch room
  void joinRoom(String roomCode, String userId, String userName) {
    if (_socket == null || !_socket!.connected) {
      print('Socket not connected, connecting first...');
      connect();
    }

    _currentRoomCode = roomCode;
    _userId = userId;

    _socket?.emit('join-room', {
      'roomCode': roomCode,
      'userId': userId,
      'userName': userName,
    });
  }

  /// Leave current room
  void leaveRoom(String roomCode) {
    _socket?.emit('leave-room', {'roomCode': roomCode});
    _currentRoomCode = null;
  }

  /// Emit video play event
  void emitPlay(double currentTime) {
    if (_currentRoomCode == null) return;
    _socket?.emit('video-play', {
      'roomCode': _currentRoomCode,
      'currentTime': currentTime,
      'userId': _userId,
    });
  }

  /// Emit video pause event
  void emitPause(double currentTime) {
    if (_currentRoomCode == null) return;
    _socket?.emit('video-pause', {
      'roomCode': _currentRoomCode,
      'currentTime': currentTime,
      'userId': _userId,
    });
  }

  /// Emit video seek event
  void emitSeek(double currentTime) {
    if (_currentRoomCode == null) return;
    _socket?.emit('video-seek', {
      'roomCode': _currentRoomCode,
      'currentTime': currentTime,
      'userId': _userId,
    });
  }

  /// Emit episode change event
  void emitEpisodeChange(int serverIndex, int episodeIndex) {
    if (_currentRoomCode == null) return;
    _socket?.emit('episode-change', {
      'roomCode': _currentRoomCode,
      'serverIndex': serverIndex,
      'episodeIndex': episodeIndex,
      'userId': _userId,
    });
  }

  /// Request current sync state from server
  void requestSync() {
    if (_currentRoomCode == null) return;
    _socket?.emit('sync-request', {'roomCode': _currentRoomCode});
  }

  /// Close room (host only)
  void closeRoom(String roomCode) {
    _socket?.emit('close-room', {'roomCode': roomCode});
    _currentRoomCode = null;
  }

  /// Dispose all stream controllers
  void dispose() {
    disconnect();
    _onVideoPlayController.close();
    _onVideoPauseController.close();
    _onVideoSeekController.close();
    _onSyncStateController.close();
    _onEpisodeChangeController.close();
    _onUserJoinedController.close();
    _onUserLeftController.close();
    _onRoomClosedController.close();
    _onConnectedController.close();
  }
}
